33 prepare for software release

rogerkuou opened the pull request:
- [x] Fix: rename `points` dimension to `space` 
- [x] Tests
   - [x] Tests for _io.py
- [x] Documentation
    - [x] add logo 
    - [x] mkdoc rendering
    - [x] Function Examples
    - [x] Add tutorial
- [x] Add a `pyproject.toml` file
    - [x] Update doc link
    - [x] Update change log link 
- [x] GitHub actions
    - [x] Build
    - [x] Doc deploy when merge to main
    - [x] Test  
- [x] Citations
    - [x] cff
    - [x] zenodo json
- [x] README
    - [x] Update description
    - [x] Installation example
    - [x] Checklist Badge:  [![OpenSSF Best Practices](https://www.bestpractices.dev/projects/7980/badge)](https://www.bestpractices.dev/projects/7980)
    - [x] Other badges:
        - [x] Sonarcloud
- [x] Update author names:
    - [x] .cff
    - [x] pyproject.toml
    - [x] zenodo.json

SarahAlidoost commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,31 @@
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
```
is there a reason why it does not run on `windows` and `mac-os`? it is possible to create a matrix for os too, see example [here](https://github.com/EcoExtreML/zampy/blob/26beebb53bda579e9bdac67fd12fbd5700165954/.github/workflows/build.yml#L19).

SarahAlidoost commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,31 @@
+      uses: actions/setup-python@v3
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install flake8 pytest pytest-cov pycodestyle
+        python -m pip install pip install .
```
I noticed that this workflow is called "**Build**", but actually, there is no section `build` here, rather the package is installed in the section "`Install dependencies`". I suggest creating a new section `-name: Build` and build the package using `python -m build` command. To do that, `build` should be installed in `Install dependencies` as `pip install build`. 
Also, I saw that `flake8 pytest pytest-cov pycodestyle` are included here as dependencies for the `Build`.  But actually these packages are not required for build but rather for testing or development. I suggest removing them form here. Therefore, the workflow can be written as:
```yaml

    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        python -m pip install build
    - name: Build
      run: |
        python -m build
```

SarahAlidoost commented on .github/workflows/doc_build.yml:
```diff
@@ -0,0 +1,28 @@
+# Build Documentation when new release created
```
To make it more clear, let's name this workflow and the file name, `doc_deploy` as it is only used when there is a release.

SarahAlidoost commented on CITATION.cff:
```diff
@@ -0,0 +1,35 @@
+    family-names: van Leijen
+identifiers:
+  - type: doi
+    value: 10.5281/zenodo.7717112
```
the version shouldn't be the newest version i.e. `10044256`? :thinking:

SarahAlidoost commented on zenodo.json:
```diff
@@ -0,0 +1,32 @@
+      "orcid": "0000-0002-5373-5209"
+    },
+    {
+      "name": "Alidoost,  Fakhereh"
```
```suggestion
      "name": "Alidoost,  Fakhereh (Sarah)"
```

SarahAlidoost commented on pyproject.toml:
```diff
@@ -0,0 +1,71 @@
+
+[project.urls]
+repository = "https://github.com/MotionbyLearning/sarxarray"
+doi = "https://doi.org/10.5281/zenodo.7717112"
```
please check zenodo version number here

SarahAlidoost commented on pyproject.toml:
```diff
@@ -0,0 +1,71 @@
+license = {file = "LICENSE"}
+authors = [
+    {name = "Ou Ku", email = "o.ku@esciencecenter.nl"},
+    {name = "Fakhereh Alidoost"},
```
```suggestion
    {name = "Fakhereh (Sarah) Alidoost"},
```

fnattino reviewed (commented):
Hi @rogerkuou, amazing! I love the docs style and really like the logo! 

Below are my comments and suggestion, I just add these as plain comments (you can see what you feel can be addressed here, what you want to leave for later or completely discard). 

Very nice work!
 

Documentation:

* Syntax highlighting is missing in the pages derived from the MD files (it works in the rendered notebook). 
* What do the various operations under `Usage` / `Common SLC operations` mean? Maybe these are all well-known tasks in the field, but maybe just some links to resources/wikipedia might help. Probably not a priority and maybe not needed for now. You could consider adding a sentence at the top of the page saying "Work in progress" or "Details coming soon ...".

Citation files:

* I would add a `title` and `abstract` fields in the CITATION.cff file, which are then translated into corresponding fields in the Zenodo config file when you do the conversion. These are important because they are the actual fields that are shown on [the Zenodo page](https://zenodo.org/records/10044256)!
* Is the webhook to make automatic Zenodo releases with GH releases activated? In principle you should have Zenodo automatically getting a snapshot from the GH repository and issueing a version DOI when you do a release.
* The Zenodo config file should be `.zenodo.json` I believe (with the leading ".").
* I see you have tagged the latest version as `2023.03`, do you plan to keep using this naming or use semantic versioning? I see that the version in the `pyproject.toml` file is going to be 0.1 - just make sure to use a consistent versioning everywhere

GH actions:

* In the "build" workflow, should you not run tests as well? Right now you are only installing the package with pytest and coverage tools (these could actually be discarded, since you are doing coverage analysis in the sonar cloud workflow, right?). Also there seem to be some issue with the last lines of the build workflow (see below).
* The action that automatically uploads the package to PyPI after a release is pretty simple (see e.g. [the one for CGC](https://github.com/phenology/cgc/blob/master/.github/workflows/pypi.yml)) and it only requires you to create an account on PyPI and getting a token to be added to the repository as a GH secret. You might consider adding that one here (so you already tackle #37) .

Information for the README file:

* I think it is considered best practice to provide information about how to contribute to the software. There are some default files around about how to contribute in GH (issues, PR, etc.), see e.g. the Python template from the eScience Center. You could then add a line to the README linking to the file: "If you want to contribute to the development of sarxarray, have a look at the [contribution guidelines]()".

SarahAlidoost commented on docs/common_ops.md:
````diff
@@ -0,0 +1,41 @@
+Compute coherence between two SLCs:
+
+```python
+slc1 = stack.isel(time=[0]) # first image
+slc2 = stack.isel(time=[2]) # third image
+window = (4,4)
+
+coherence = slc1.slcstack.complex_coherence(slc2, window)
````
```suggestion
from sarxarray import complex_coherence
slc1 = stack.complex.isel(time=0) # first image
slc2 = stack.complex.isel(time=2) # third image
window = (4,4)

coherence = complex_coherence(slc1, slc2, window)
```

SarahAlidoost commented on docs/data_loading.md:
````diff
@@ -0,0 +1,64 @@
+stack_smallchunk = sarxarray.from_binary(list_slcs, shape, chunks=(2000, 2000))
+```
+
+
+
+
+
+
+
+
+
+
````
```suggestion
```

SarahAlidoost commented on docs/index.md:
```diff
@@ -0,0 +1,23 @@
+
+All the above functionalities can be scaled up to a Hyper-Performance Computation (HPC) system.
+
+
+
+
+
+
+
```
```suggestion
```

SarahAlidoost commented on docs/setup.md:
```diff
@@ -0,0 +1,21 @@
+
+## Tips
+
+We strongly recommend installing separately from your default Python envrionment. E.g. you can use enviroment manager e.g. [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/download.html) to create separate environment.
```
I suggest refering users to mamba instead of conda https://mamba.readthedocs.io/en/latest/mamba-installation.html

SarahAlidoost reviewed (changes requested):
@rogerkuou well done! very nice and valuable additions to the package :clap: . Please see my comments/suggestions. If something is unclear, let me know. 
One more suggestion: we can also add a `contributing.md` or `developer guide` to the documentation to show others how to for example, run tests or build the documentation locally. This can be added either in this pull request or in a different one. I can help with creating it too.

SarahAlidoost commented on pyproject.toml:
```diff
@@ -0,0 +1,71 @@
+repository = "https://github.com/MotionbyLearning/sarxarray"
+doi = "https://doi.org/10.5281/zenodo.7717112"
+documentation = "https://motionbylearning.github.io/sarxarray/"
+changelog = "https://motionbylearning.github.io/sarxarray/CHANGLOG/"
```
if file `CHANGLOG` is renamed, please fix it here too

SarahAlidoost commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,39 @@
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install build
+        python -m install
```
```suggestion
        python -m install .
```

rogerkuou commented:
Hi @fnattino and @SarahAlidoost, thanks for the review! I made update accoding to your comments, below is a summary:

**GH actions**:
- Added build tests in `build.yml`
- Fixed dependency installation in `build.yml`. Now the package dependencies are not installed in `build.yml`. That is down in sonarcloud.yml
- Add Doc build test in `build.yml`
- I still did not run unit test in `build.yml` since I think it is already covered in `sonacloud.yml`
- Added `pipy.yml`. It will be triggered by a release. I will keep #37 open for now.

**Docs**
- Added some mkdocs extensions and enabled the python syntax rendering
- Add a WIP claim in the "common operations" Tab. Submitted an [issue](https://github.com/MotionbyLearning/sarxarray/issues/38) for future work.
- Fixed type in CHANGELOG.md and updated links

**Citation**
- Updated the DOI in README to [10.5281/zenodo.7717027](https://zenodo.org/doi/10.5281/zenodo.7717027), it refers to all versions now.
- Updated zenodo name to .zenodo.json
- Applied Sarah's suggestion on her name >_-

Could you please have another look?

fnattino commented:
Great @rogerkuou!

> * I still did not run unit test in `build.yml` since I think it is already covered in `sonacloud.yml`

The only difference is that build runs for a set of OSs and Python versions, while sonar cloud can run for a single OS and version. So if you would like to find issues with different setups, you can run tests (without coverage) in build as well.

The only missing thing, in case you want to add it, is the contributing instructions. But for the rest, looks very good to me!

fnattino commented on docs/data_loading.md:
````diff
@@ -0,0 +1,55 @@
+```
+You can also skip the `dtype` argument since it's defaulted to `np.complex64`. The stack will be read as an `xarray.Dataset` object, with data variables lazily loaded as `Dask Array`:
+
+```output
+print(stack)
+
+<xarray.Dataset>
+Dimensions:    (azimuth: 10018, range: 68656, time: 3)
+Coordinates:
+  * azimuth    (azimuth) int64 0 1 2 3 4 5 ... 10013 10014 10015 10016 10017
+  * range      (range) int64 0 1 2 3 4 5 ... 68650 68651 68652 68653 68654 68655
+  * time       (time) int64 0 1 2
+Data variables:
+    complex    (azimuth, range, time) complex64 dask.array<chunksize=(4000, 4000, 1), meta=np.ndarray>
+    amplitude  (azimuth, range, time) float32 dask.array<chunksize=(4000, 4000, 1), meta=np.ndarray>
+    phase      (azimuth, range, time) float32 dask.array<chunksize=(4000, 4000, 1), meta=np.ndarray>
+```
````
Just noticed this (very) minor formatting issue:

````suggestion
```python
print(stack)
```

```output
<xarray.Dataset>
Dimensions:    (azimuth: 10018, range: 68656, time: 3)
Coordinates:
  * azimuth    (azimuth) int64 0 1 2 3 4 5 ... 10013 10014 10015 10016 10017
  * range      (range) int64 0 1 2 3 4 5 ... 68650 68651 68652 68653 68654 68655
  * time       (time) int64 0 1 2
Data variables:
    complex    (azimuth, range, time) complex64 dask.array<chunksize=(4000, 4000, 1), meta=np.ndarray>
    amplitude  (azimuth, range, time) float32 dask.array<chunksize=(4000, 4000, 1), meta=np.ndarray>
    phase      (azimuth, range, time) float32 dask.array<chunksize=(4000, 4000, 1), meta=np.ndarray>
```
````

rogerkuou commented:
> . But for the rest, looks very good to me!

Ah thanks, I will apply both: 1) add a tests in build, and 2) add a contributing instruction

rogerkuou commented:
Hi @fnattino, I added the tests to `build.yml` and also added Contribution doc (mainly adpated from the CGC package). Would you like to give another look?

fnattino reviewed (approved):
All looks good to me! :)

SarahAlidoost commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,65 @@
+      fail-fast: false
+      matrix:
+        os: ['ubuntu-latest', 'macos-latest', 'windows-latest']
+        python-version: ["3.10", "3.11", "3.12"]
```
There is no python version `3.12` in `pyproject.toml` file. Do we want to support it? if so, it should be added to the `pyproject.toml `, otherwise let's remove it form here.

rogerkuou replied:
I will added it to `pyproject.toml`

SarahAlidoost commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,65 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        os: ['ubuntu-latest', 'macos-latest', 'windows-latest']
```
For testing the documentation, one `os` e.g. `ubuntu-latest` should be sufficient, we can remove the rest to make it concise and fast.

SarahAlidoost commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,65 @@
+      fail-fast: false
+      matrix:
+        os: ['ubuntu-latest', 'macos-latest', 'windows-latest']
+        python-version: ["3.10", "3.11", "3.12"]    
```
see my comment about `3.12` above

SarahAlidoost commented on .github/workflows/pypi.yml:
```diff
@@ -0,0 +1,25 @@
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install build twine
```
```suggestion
        python -m pip install build 
```

SarahAlidoost replied:
`pypa/gh-action-pypi-publish` uses `twine` internally, no need to install it.

SarahAlidoost commented on .github/workflows/pypi.yml:
```diff
@@ -0,0 +1,25 @@
+        python -m pip install build twine
+        python -m build
+    - name: Publish package
+      uses: pypa/gh-action-pypi-publish@master
```
Why do you use `@master` version? :thinking: 
in this [documentation](https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python#publishing-to-package-registries), the `@v1` is used. In the same documentation, they refer to an example [python-publish](https://github.com/actions/starter-workflows/blob/main/ci/python-publish.yml) where `@27b31702a0e7fc50959f5ad993c78deac1bdfc29` is used.

rogerkuou replied:
you are right, on their [GH page](https://github.com/pypa/gh-action-pypi-publish) the `release/v1` version is used. Will do that one.

SarahAlidoost commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,44 @@
+- (**important**) announce your plan to the rest of the community *before you start working*. This announcement should be in the form of a (new) issue;
+- (**important**) wait until some kind of consensus is reached about your idea being a good idea;
+- if needed, fork the repository to your own Github profile and create your own feature branch off of the latest master commit. While working on your feature branch, make sure to stay up to date with the master branch by pulling in changes, possibly from the 'upstream' repository (follow the instructions [here](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and [here](https:// ...
+- make sure the existing tests still work by running, e.g. `pytest tests`;
```
```suggestion
- make sure the existing tests still work. First, install the development dependencies as `pip install .[dev]`, and then run `pytest tests`;
```

SarahAlidoost commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,44 @@
+- if needed, fork the repository to your own Github profile and create your own feature branch off of the latest master commit. While working on your feature branch, make sure to stay up to date with the master branch by pulling in changes, possibly from the 'upstream' repository (follow the instructions [here](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and [here](https:// ...
+- make sure the existing tests still work by running, e.g. `pytest tests`;
+- add your own tests (if necessary);
+- update or expand the documentation;
```
```suggestion
- update or expand the documentation. Make sure the documentation is built successfully: first, install documentation dependencies as `pip install .[docs]` and then run `mkdocs build`.
```

SarahAlidoost reviewed (approved):
@rogerkuou thanks for addressing the comments. I added some suggestions/comments and approved this PR.
